README.md
# metempsychose
Game by Cavalcadeur
http://cavalcadeur.github.io/metempsychose/


changelog.txt
--- objectifs 0.8 ---
- plus de carte parce qu'il faut repenser le système   [DONE]
- niveaux aquatiques 0/5
- boss de fin et clés                                  [DONE]
- bonus après générique
- générique                                            [DONE]
- animation d'entrée des portes                        [DONE]
- optimisation hiérarchique des différentes animations [DONE] (parce que le jeu en lui même était enterré sous les conditions)
- meilleur gestion des boules de feu.                  [DONE]

*** version 0.7 ***
- 15 niveaux en plus
- 6 nouveaux persos: mage blanc, mage vert, champi ninja, champi vortex, sauterelle, torche
- changement de famille pour la bombe
- menu tceles
- cinématiques
- optimisation
- changement du dessin des plates-formes

*** version 0.6 ***
- nouveaux persos : Mage violet, Mage noir, GraviChampi
- Création d'un éditeur
- Correction de bugs majeurs
- Beaucoup plus de niveaux (10 ---> 40)
- niveaux secrets
- amélioration de l'aide
- ajout de Crossed pour le contenu additionnel

*** version 0.5 ***
- Niveaux à l'aveugle
- plus de niveaux d'adresse
- ajout de la musique en fond
- modification du fonctionnement des interrupteurs

*** version 0.4 ***
- ajout du monde 3
- ajout du fantôme
- création de l'aide
- décors
- animation de l'explosion de la bombe

*** version 0.3 ***
- Fumée pour la bombe
- plus de niveau
- nouveau fond d'écran
- nouveaux persos : boule de neige et Champi géant

*** version 0.2 ***
- ajout de persos : plante carnivore, Mage Electro
- Menu
- portes
- niveau 1-2 disponible et quelques niveaux d'adresse

*** version 0.1 ***
- Un niveau
- 6 persos : Mage Feu, Boule, Champique, Jumper, Psychos, Bombe
- système de metempsychose au point
- interrupteurs


test.js
function start() {
    var body = document.body;
    var canvas = document.getElementById('canvas');
    canvas.setAttribute('width', 3000);
    canvas.setAttribute('height', 2000);
    canvas.style.width = '3000px';
    canvas.style.height = '2000px';
    
    var ctx = canvas.getContext("2d");
    for (var i=0 ; i < 1000 ; i++) {
        ctx.fillStyle = "rgb(" + Math.floor(256 * Math.random()) + "," + Math.floor(256 * Math.random()) + "," + Math.floor(256 * Math.random()) + ")";
        ctx.fillRect(Math.random() * 3000, Math.random() * 2000, 30, 30);
    }

    var keys = {};
    document.addEventListener(
        "keydown",
        function (event){
            event.preventDefault();
            event.stopPropagation();
            keys[event.keyCode] = 1;
        }
    );
    document.addEventListener(
        "keyup",
        function (event){
            event.preventDefault();
            event.stopPropagation();
            keys[event.keyCode] = 0;
        }
    );

    var paint = function() {
        if (keys[39]) {
            body.scrollLeft += 5;
        }
        if (keys[40]) {
            body.scrollTop += 5;
        }
        requestAnimationFrame(paint);
    };

    paint();
}


nuage.js
var newExplosion = function(x,y,tx,ty,s,color,n){
    var Widget = require("wdg");
    var time = 4;
    var random = 1;
    var v = 0;
    var colorset = [["rgba(229,179,24,1);rgba(150,35,38,1);rgba(150,35,38,0)","rgba(216,140,28,1);rgba(119,35,40,1);rgba(119,35,40,0)"],["rgba(50,50,50,1);rgba(0,0,0,1);rgba(0,0,0,0)","rgba(255,255,255,1);rgba(200,200,200,1);rgba(200,200,200,0)"],["rgba(107,30,136,1);rgba(42,12,54,1);rgba(42,12,54,0)","rgba(32,153,203,1);rgba(20,98,130,1);rgba(20,98,130,0)"],["rgba(107,30,136,1);rgba(42,12,54,1);rgba(42,12,54,0)","rgba(32,153,203,1);rgba(20,98,130,1);rgba(20,98,130,0)"]];

    var square = new Widget({id: "square"});

    var h = Widget.svg({
        preserveAspectRatio: 'xMidYMid',
        width: s,
        height: s,
        viewBox: "0 0 100 100"
    });
    h.addClass("nuage");

    h.append(
        Widget.svg('g',{
            stroke: "none",
            "stroke-width": 0.1,
            opacity: 1,
            transform: "rotate(360, 50, 50)"
        }).append(
            Widget.svg('animateTransform', {
                attributeName: "transform",
                attributeType: "XML",
                type: "rotate",
                values: "0,50,50;-"+rnd(3)+3+",50,50",
                dur: time,
                repeatCount: "1"
            }), 
            Widget.svg('path', {
                d: "M50,20C60,10,100,30,85,50C100,60,80,100,45,80C40,87,-5,80,10,50C-10,30,30,0,50,20",
                fill: "rgba(229,179,24,0)"
            }).append(
                Widget.svg('animate', {
                    attributeName: "fill",
                    dur: time,
                    repeatCount: "1",
                    values: colorset[color][0]
                }),
                Widget.svg('animate', {
                    attributeName: "d",
                    dur: time,
                    repeatCount: "1",
                    values: "M50,20C60,10,100,30,85,50C100,60,80,100,45,80C40,87,-5,80,10,50C-10,30,30,0,50,20;M40,17C50,0,90,20,75,49C84,77,66,83,46,71C40,87,-5,80,10,50C-10,30,30,0,40,17;M40,17C50,0,90,20,75,49C84,77,66,83,46,71C40,87,-5,80,10,50C-10,30,30,0,40,17"
                })) ,
            Widget.svg('path', {
                d: "M70,40C63,48,45,30,53,22C60,12,98,32,80,52C100,60,70,98,44,77C40,82,3,80,13,50C-5,30,25,5,50,20C40,30,60,47,70,40",
                fill: "rgba(216,140,28,0)"
            }).append(
                Widget.svg('animate', {
                    attributeName: "fill",
                    dur: time,
                    repeatCount: "1",
                    values: colorset[color][1]
                }),
                Widget.svg('animate', {
                    attributeName: "d",
                    dur: time,
                    repeatCount: "1",
                    values: "M70,40C63,48,45,30,53,22C60,12,98,32,80,52C100,60,70,98,46,76C40,82,3,80,13,50C-5,30,25,5,50,20C40,30,60,47,70,40C70,40,70,40,70,40;M60,30C67,48,35,29,43,21C50,2,88,22,70,48C77,70,60,78,46,66C40,82,3,80,13,50C-5,30,20,8,40,17C30,27,60,47,63,37C64,33,61,31,60,30;M60,30C67,48,35,29,43,21C50,2,88,22,70,48C77,70,60,78,46,66C40,82,3,80,13,50C-5,30,20,8,40,17C30,27,60,47,63,37C64,33,61,31,60,30"
                }))
        )
    ).css({
        position: "absolute",
        left: x,
        top: y
    });
    h.rect();
    window.setTimeout(function () {
        h.css({
            left: tx,
            top: ty
        });
    });
    console.info("[nuage] h.css('left')=...", h.css('left'));
    var timeOut = window.setTimeout(function(){
        h.detach();
    },time * 1000);
    square.append(h);
}


weird.js
function wrongLevelS (){
    actor[0].moves.img = "E";
    actor[0].moves.sy = 107;
    actor[1].moves.img = "F3";
    actor[1].moves.sx = 103;
    actor[1].moves.sy = 103;
    actor[2].moves.img = "champiDiamant";
    actor[2].moves.sx = 70;
    actor[2].moves.sy = 60;
    actor[3].moves.img = "barre2";
    actor[3].moves.sx = 50;
    actor[3].moves.sy = 50;
    actor[4].moves.img = "inter";
    actor[4].moves.sx = 57;
    actor[4].moves.sy = 59;
    actor[5].moves.img = "esprit";
    actor[5].moves.sx = 100;
    actor[5].moves.sy = 100;
    actor[6].moves.img = "T";
    actor[6].moves.sx = 100;
    actor[6].moves.sy = 107;
    actor[7].moves.img = "champiNinja";
    actor[7].moves.sx = 100;
    actor[7].moves.sy = 107;
    actor[8].moves.img = "bombe3";
    actor[8].moves.sx = 60;
    actor[8].moves.sy = 83;
    actor[9].moves.img = "fantome";
    actor[9].moves.sx = 50;
    actor[9].moves.sy = 105;
    actor[10].moves.img = "mageNoir";
    actor[10].moves.sx = 113;
    actor[10].moves.sy = 155;
    actor[13].moves.img = "note1";
    actor[13].moves.sx = 15;
    actor[13].moves.sy = 26;
    actor[11].moves.img = "heroine2";
    actor[11].moves.sx = 150;
    actor[11].moves.sy = 200;
}


Crossed.js
var Crossed = function () {
    var menu = 0;
    var passWord = [0,6,6,6,0];
    var tryWord = [0,0,0,0,0];
    var nWord = 0;
    var tWord = 0;
    var ecart,taille,marge;
    function drawNumber(x,y,n,ctx){
	ctx.fillStyle = "rgb(200,200,200)";
        ctx.strokeStyle = "rgb(200,200,200)";
        ctx.lineWidth = 5;
	if (n == 1 | n == 6){ 
            ctx.beginPath();
            ctx.moveTo(x + taille/4,y + taille/4);
            ctx.lineTo(x + taille/4*3,y + taille/4);
            ctx.lineTo(x + taille/4*3,y + taille/4*3);
            ctx.lineTo(x + taille/4,y + taille/4*3);
            ctx.lineTo(x + taille/4,y + taille/4);
        }
        else if (n == 2 | n == 7){ 
            ctx.beginPath();
            ctx.moveTo(x + taille/2,y + taille/4);
            ctx.lineTo(x + taille/4*3,y + taille/4*3);
            ctx.lineTo(x + taille/4,y + taille/4*3);
            ctx.lineTo(x + taille/2,y + taille/4);
        }
        else if (n == 0 | n == 5){ 
            ctx.beginPath();
            ctx.arc(x + taille/2,y + taille/2,taille / 4,-Math.PI,Math.PI);
        }
        else if (n == 3 | n == 8){ 
            ctx.beginPath();
            ctx.arc(x + taille/2,y + taille/2,taille / 4,0,Math.PI);
        }
        else if (n == 4 | n == 9){ 
            ctx.beginPath();
            ctx.moveTo(x + taille/2,y + taille/6);
            ctx.lineTo(x + taille/4*3,y + taille/3);
            ctx.lineTo(x + taille/4*3,y + taille/3*2);
            ctx.lineTo(x + taille/2,y + taille/6*5);
            ctx.lineTo(x + taille/4,y + taille/3*2);
            ctx.lineTo(x + taille/4,y + taille/3);
        }
        ctx.closePath();
        if (n < 5) ctx.stroke();
        else ctx.fill();
    }

    function contenu(){
        var N = tryWord[4] + tryWord[3]*10 + tryWord[2]*100 + tryWord[1]*1000 + tryWord[0]*10000;
        if (N == 25652) selection("martheo1");
        else if (N == 19783) selection("maker1");
        else if (N == 03778) selection("secretStage");
    }

    return {
	drawMenu: function( ctx , W, H ) {
	    ctx.globalAlpha = 0.5;
	    ctx.fillStyle = "rgb(200,200,200)";
	    ctx.fillRect(0,0,W,H);
	    passWord.forEach(
		function(e,i){
		    ctx.fillStyle = "rgb(0,0,0)";
		    ctx.strokeStyle = "rgb(250,250,250)";
		    ctx.lineWidth = 10;
		    ctx.fillRect(ecart*(i+1) + taille*i,marge,taille,taille);
		    if (i < nWord) drawNumber(ecart*(i+1) + taille*i,marge,e,ctx);
		}
	    );
	    tryWord.forEach(
		function(e,i){
		    ctx.fillStyle = "rgb(0,0,0)";
		    ctx.strokeStyle = "rgb(250,250,250)";
		    ctx.lineWidth = 10;
		    ctx.fillRect(ecart*(i+1) + taille*i,marge*5 + taille,taille,taille);
		    if (i == tWord) ctx.strokeRect(ecart*(i+1) + taille*i,marge*5 + taille,taille,taille);
		    drawNumber(ecart*(i+1) + taille*i,marge*5 + taille,e,ctx);
		}
	    );
	    ctx.globalAlpha = 1;
	    animation();
	},

	keyCrossed: function( ) {
	    if (menu == 0) { menu = 1; goOn = 0;}
	    else {menu = 0;goOn = 1;contenu();}	    
	},

	testCrossed: function( ) {
	    if (menu == 0) return 0;
	    else return 1;	    
	},

	init: function( W, H ) {
	    ecart = W/18;
	    taille = (W - ecart*6) / 5;
	    marge = (H - taille*2) / 6;
	},

        deplacement: function(n){
            if (n == 2) tryWord[tWord] = (tryWord[tWord]+1)%10;
            else if (n == -2) tryWord[tWord] = (tryWord[tWord]-1)%10;
            else if (n == 1) tWord += 1;
            else if (n == -1) tWord -= 1;
            if (tryWord[tWord] == -1) tryWord[tWord] = 9;
            if (tWord >= 5) tWord = 4;
            else if (tWord < 0) tWord = 0;
        },

        improve: function(){
            if (nWord < 5) nWord += 1;

        },
        
        keysPress: function(keyCode){
            if (keyCode == 67) Crossed.keyCrossed();
           
            if (menu == 1){
                if (keyCode == 39) Crossed.deplacement(1);
                else if (keyCode == 37) Crossed.deplacement(-1);
                else if (keyCode == 38) Crossed.deplacement(2);
                else if (keyCode == 40) Crossed.deplacement(-2);
            }

        }
    };
}();


dev.js
var W,H;
var ctx,canvas;
var X = 0;
var Y = 400;
var ep = 50;
var j = 0;
var keys = [];
var metampsy = 0;
var imgFond = new Image();
imgFond.src = "images/fond.png";
var imgFeu = new Image();
imgFeu.src = "images/feu.png";
var imgBalle = new Image();
imgBalle.src = "images/psychosBalle.png";
var imgPancarte = new Image();
imgPancarte.src = "images/pancarte.png";
var imgFeuX = 70;
var imgFeuY = 35;
var imgBalleX = 35;
var imgBalleY = 35;
var cible;
var filin = 0.05;
var filinVit = 0.002;
var mort = 0;
var vj;
var trans = 0;
var laserPsy = -10;

imgFond.onload = function (){
    console.log("Ah");
};

function Bombe() {
    this.saut = 0;
    this.vit = 3;
    this.img = "bombe";
    this.sx = 35;
    this.sy = 40;
    this.capa = "explode";
    this.IA = "wait";
    this.att = [];
    this.res = [3];
    this.explode = 0;
}

function MageFeu() {
    this.saut = 0;
    this.vit = 3;
    this.img = "F";
    this.sx = 52;
    this.sy = 91;
    this.capa = "feu";
    this.IA = "spamFeu";
    this.att = [];
    this.res = [];
    this.spam = 500;
}

function Boule() {
    this.saut = 27;
    this.vit = 3;
    this.img = "boule";
    this.sx = 63;
    this.sy = 63;
    this.capa = "instable";
    this.inertie = 0;
    this.r = 0;
    this.IA = "wait";
    this.att = [2,1];
    this.res = [];
}

function Champique(){
    this.saut = 0;
    this.vit = 3;
    this.img = "champique";
    this.sx = 62;
    this.sy = 43;
    this.capa = "";
    this.IA = "allerRetour";
    this.att = [4];
    this.res = [1,2,4];
}

function Champi(){
    this.saut = 0;
    this.vit = 3;
    this.img = "champi";
    this.sx = 50;
    this.sy = 48;
    this.capa = "";
    this.IA = "allerRetour";
    this.att = [1];
    this.res = [];
}

function Jumper(){
    this.saut = 20;
    this.vit = 0;
    this.img = "jumper";
    this.sx = 60;
    this.sy = 56;
    this.capa = "";
    this.IA = "simpleJump";
    this.att = [2,4];
    this.res = [1,2,3,4,5,6,7,9];
}

function Psychos() {
    this.saut = 0;
    this.vit = 0;
    this.img = "psychos";
    this.sx = 46;
    this.sy = 40;
    this.capa = "laserPsy";
    this.IA = "tirer";
    this.att = [8];
    this.res = [8];
}

function Boxe(){
    this.saut = 0;
    this.vit = 0;
    this.img = "boxe";
    this.sx = 50;
    this.sy = 50;
    this.capa = "antipsy";
    this.IA = "wait";
    this.att = [1];
    this.res = [1,2,3,4,5,6,7,8,9];
}

function MageElectro(){
    this.saut = 0;
    this.vit = 15;
    this.img = "E";
    this.sx = 57;
    this.sy = 91;
    this.capa = "courseLongue";
    this.IA = "wait";
    this.att = [7];
    this.res = [7];

}

function Carnivore() {
    this.saut = 0;
    this.vit = 0;
    this.img = "carnivore";
    this.sx = 57;
    this.sy = 66;
    this.capa = "feu";
    this.IA = "spamFeu";
    this.att = [7];
    this.res = [1,4,5,6,7,8,9];
    this.spam = 500;
}

function Esprit() {
    this.saut = 5;
    this.vit = 6;
    this.img = "fantome";
    this.sx = 32;
    this.sy = 67;
    this.capa = "immateriel";
    this.IA = "maintien";
    this.att = [];
    this.res = [];
    this.spam = 80;
}

function Interrupteur(){
    this.saut = 0;
    this.vit = 0;
    this.img = "inter";
    this.sx = 57;
    this.sy = 59;
    this.capa = "interrupteur";
    this.IA = "wait";
    this.att = [];
    this.res = [1,3,4,5,6,7,8,9];
    this.mode = 0;
}

var element = {"feu":[],"balle":[],"panneau":[]};

// niveau

var limite = 2000;
var niveau = [[1900+ep,250,500,ep],[2400,0,ep,250+ep],[2400,249+ep,ep,250+ep],[1700+ep,500+ep,700,ep],[1250,250,250,ep],[1250,100,ep,150],[900,250,300,ep],[800,450,100,ep],[900,650,100,ep],[800,850,100,ep],[900,1050,100,ep],[200,1250,700,ep],[200,850,550,ep],[100,250,100,ep],[0,450,100,ep],[100,650,100,ep],[0,850,100,ep],[100,1050,100,ep],[0-ep,100,ep+1,1150+ep],[0-ep,1250+ep,300,ep],[1200,1050,1000,ep],[1200,750,ep,300],[-300,1500,600,ep*3]];

element.balle = [[1700,500],[133,215],[1300,1400]];
var balles = 2;
element.panneau = [[2000,500+ep,"Les deux mages doivent parvenir en bas en prenant le mage vert au passage."]];

var actor = [{"x":2000,"y":100,"vx":0,"vy":0,"g":0,"saut":0,"moves":new MageFeu,"img":new Image()},
             {"x":2300,"y":100,"vx":0,"vy":0,"g":0,"saut":0,"moves":new MageElectro,"img":new Image()},
             {"x":1100,"y":100,"vx":0,"vy":0,"g":0,"saut":0,"moves":new Boule,"img":new Image()},
             {"x":2300,"y":550,"vx":0,"vy":0,"g":0,"saut":0,"moves":new Interrupteur,"img":new Image(),"plate":[1499,250,402+ep,ep]},
             {"x":500,"y":300,"vx":0,"vy":0,"g":0,"saut":0,"moves":new Psychos,"img":new Image()},
             {"x":1300,"y":1000,"vx":0,"vy":0,"g":0,"saut":0,"moves":new Carnivore,"img":new Image()},
             {"x":-250,"y":2000,"vx":0,"vy":0,"g":0,"saut":0,"moves":new Esprit,"img":new Image()}];
actor[3].moves.res = [1,4,5,6,7,8,9];
// programme

function rnd(max){
    return Math.floor(Math.floor(Math.random()*max));
}

function tombe(n){
    if (actor[n].moves.capa == "interrupteur")return false;
    if (actor[n].moves.capa == "immateriel")return true;
    var sautG = 0.2;
    niveau.forEach(
        function(c) {
            if (c[1] < actor[n].y - actor[n].moves.sy && c[1] + c[3] > actor[n].y - actor[n].moves.sy && c[0] < actor[n].x && c[0] + c[2] > actor[n].x){
                if (actor[n].g < 0) actor[n].g = 0;
                actor[n].y = c[1] + c[3] + 1 + actor[n].moves.sy;
            }
            if (c[1] < actor[n].y && c[1] + c[3] > actor[n].y && c[0] < actor[n].x && c[0] + c[2] > actor[n].x){
                if (actor[n].g > 0) actor[n].g = 0;
                actor[n].y = c[1];
                if (actor[n].saut != 0 && actor[n].g == 0)actor[n].saut -= 1;
                sautG = 0;
                return;
            }
        }
    );
    if (actor[n].moves.capa == "instable"){
        if (actor[n].moves.inertie > 0)actor[n].moves.r += sautG;
        if (actor[n].moves.inertie < 0)actor[n].moves.r -= sautG;
        if (actor[n].moves.r < Math.PI)actor[n].moves.r += 2*Math.PI;
        if (actor[n].moves.r > Math.PI)actor[n].moves.r -= 2*Math.PI;
    }
    return true;
}

function droite(n){
    if (actor[n].moves.capa == "immateriel")return;
    niveau.forEach(
        function(c) {
            if (c[1] - 1 < actor[n].y - actor[n].moves.sy / 2 && c[1] + c[3] > actor[n].y - actor[n].moves.sy / 2 && c[0] < actor[n].x + actor[n].moves.sx / 2 && c[0] + c[2] > actor[n].x + actor[n].moves.sx / 2){
                actor[n].x = c[0] - 1 - actor[n].moves.sx / 2;
                if (actor[n].moves.capa == "courseLongue") actor[n].vx = 0;
                if (actor[n].moves.capa == "instable") actor[n].moves.inertie = 0;
            }
        }
    );
}

function gauche(n){
    if (actor[n].moves.capa == "immateriel")return;
    niveau.forEach(
        function(c) {
            if  (c[1] - 1 < actor[n].y - actor[n].moves.sy / 2 && c[1] + c[3] > actor[n].y - actor[n].moves.sy / 2 && c[0] < actor[n].x - actor[n].moves.sx / 2 && c[0] + c[2] > actor[n].x - actor[n].moves.sx / 2){
                actor[n].x = c[0] + c[2] + 1 + actor[n].moves.sx / 2;
                if (actor[n].moves.capa == "courseLongue") actor[n].vx = 0;
                if (actor[n].moves.capa == "instable") actor[n].moves.inertie = 0;
            }
        }
    );
}

function moveRight(n){
    actor[n].vx += actor[n].moves.vit;
    if (actor[n].moves.capa == "courseLongue") actor[n].vx = actor[n].moves.vit;
    if (actor[n].moves.capa == "instable"){
        actor[n].moves.inertie += 0.1;
        actor[n].moves.r += actor[n].moves.vit / actor[n].moves.sy;
        if (actor[n].moves.r > Math.PI)actor[n].moves.r -= 2*Math.PI;
    }

}

function moveLeft(n){
    actor[n].vx -= actor[n].moves.vit;
    if (actor[n].moves.capa == "courseLongue") actor[n].vx = -1 * actor[n].moves.vit;
    if (actor[n].moves.capa == "instable"){
        actor[n].moves.inertie -= 0.1;
        actor[n].moves.r -= actor[n].moves.vit / actor[n].moves.sy;
        if (actor[n].moves.r < Math.PI)actor[n].moves.r += 2*Math.PI;
    }
}

function proche(){
    var dist = 600;
    var truc = j;
    for (var i = 0;i < actor.length;i++){
        if (Math.hypot(actor[i].x-actor[j].x , actor[i].y-actor[j].y) < dist && Math.hypot(actor[i].x-actor[j].x , actor[i].y-actor[j].y) != 0 && actor[i].moves.capa != "antipsy" && actor[i].moves.capa != "interrupteur"){dist = Math.hypot(actor[i].x-actor[j].x , actor[i].y-actor[j].y);truc = i;}
    }
    return truc;
}

function IA(i){
    if (actor[i].moves.IA == "simpleJump") {
        if (actor[i].g == 1 && actor[i].saut == 0){
            actor[i].g = -1 * actor[i].moves.saut;
            actor[i].saut = 2;
        }
    }
    else if (actor[i].moves.IA == "spamFeu") {
        if (actor[i].moves.spam > 500){
            actor[i].moves.spam = 0;
            element.feu.push([actor[i].x,actor[i].y - actor[i].moves.sy / 2,10]);
        }
    }
    else if (actor[i].moves.IA == "maintien") {
        if (actor[i].moves.spam > 100){
            actor[i].moves.spam = 0;
            actor[i].g = -1 * actor[i].moves.saut;
        }
    }
    else if (actor[i].moves.IA == "tirer" && actor[i].moves.capa == "laserPsy") {
        if (Math.abs(actor[i].x - actor[j].x) < actor[i].moves.sx / 2 + actor[j].moves.sx / 2 + 150 && Math.abs(actor[i].y - actor[i].moves.sy / 2 - (actor[j].y - actor[j].moves.sy / 2)) < actor[i].moves.sy / 2 + actor[j].moves.sy / 2 + 150){
            laserPsy = i;
            coup(j,8);
        }
    }
    contact(i,j);
}

function transfert(){
    balles -= 1;
    keys[96] = 0;
    vj = proche();
    if (vj == j) return;
    trans = 1;
}

function death(){
    mort = 1;
    actor[j].g = 10;
}

function coup(n,att){
    if (actor[n].moves.res.indexOf(att) == -1){
        if (actor[n].moves.capa != "explode" && actor[n].moves.capa != "interrupteur"){
            if (j == n)death();
            else {
                actor.splice(n,1);
                if (j > n) j -= 1;
            }
        }
        else if (actor[n].moves.capa == "interrupteur"){
            actor[n].moves.mode += 1;
            if (actor[n].moves.mode == 1)niveau.push(actor[n].plate);
            else {
                actor[n].moves.mode = 0;
                niveau.splice(niveau.indexOf(actor[n].plate),1);
            }
        }
        else {actor[n].img.src = "images/" + actor[n].moves.img + "2.png";actor[n].moves.explode = 1;}
    }
}

function contact(i,k){
    var array = actor[i].moves.att;
    if (array.length != 0){
        if (array.indexOf(2) != -1 && actor[i].x > actor[k].x - actor[k].moves.sx / 2 && actor[i].x < actor[k].x + actor[k].moves.sx / 2 && actor[i].y > actor[k].y - actor[k].moves.sy && actor[i].y < actor[k].y - actor[k].moves.sy / 2){
            coup(k,2);
        }
        else {
            if (Math.abs(actor[i].x - actor[k].x) < actor[i].moves.sx / 2 + actor[k].moves.sx / 2 && Math.abs(actor[i].y - actor[i].moves.sy / 2 - (actor[k].y - actor[k].moves.sy / 2)) < actor[i].moves.sy / 2 + actor[k].moves.sy / 2) coup(k,array[array.length - 1]);
        }
    }
}

function action(){
    element.panneau.forEach(
        function(e,index) {
            if (Math.hypot(actor[j].x - e[0],actor[j].y - e[1]) < 20){
                alert (e[2]);
                element.panneau.splice(index,1);
                actor[j].vx = 0;
            }
        }
    );
    if (actor[j].moves.capa == "immateriel")  actor[j].g = -1 * actor[j].moves.saut;
    if (actor[j].moves.capa != "feu" && actor[j].g == 1 && actor[j].saut == 0){
        actor[j].g = -1 * actor[j].moves.saut;
        actor[j].saut = 2;
    }
    if (actor[j].moves.capa == "feu" && actor[j].moves.spam > 100){
        actor[j].moves.spam = 0;
        keys[32] = 0;
        element.feu.push([actor[j].x,actor[j].y - actor[j].moves.sy / 2,10]);
    }
}

function start(){
    canvas = document.querySelector("#canvas");
    ctx = canvas.getContext("2d");
    W = canvas.width;
    H = canvas.height;
    actor.forEach(
        function(c) {
            c.img.src = "images/" + c.moves.img + ".png";
        }
    );
    document.addEventListener(
        "keydown",
        function (event){
            event.preventDefault();
            keys[event.keyCode] = 1;
        }
    );
    document.addEventListener(
        "keyup",
        function (event){
            event.preventDefault();
            keys[event.keyCode] = 0;
        }
    );
    canvas.addEventListener("click",function(evt) {
                                evt.stopPropagation();
                                evt.preventDefault();
                                //evt = evt.changedTouches[0];
                                var rect = canvas.getBoundingClientRect();
                                var x = evt.screenX - rect.left;
                                var y = evt.screenY - rect.top;
                                click(x, y);
                            });
    animation();
}

function animation(){
    var f = function(t) {
        if (trans == 0){
            if (mort != 1){
                paint(t);
                window.requestAnimationFrame(f);
            }
            else window.requestAnimationFrame(drawDeath);
        }
        else window.requestAnimationFrame(drawTransfert);
    };
    window.requestAnimationFrame(f);
}

function paint(t){
    filin += filinVit;
    if (filin > 0.2 | filin < 0.05) filinVit = filinVit * -1;
    if (1 == keys[39]) moveRight(j);
    if (1 == keys[37]) moveLeft(j);
    if ((1 == keys[96] | 1 == keys[88]) && balles > 0) transfert();
    if (1 == keys[32]) action();
    for (var i = 0;i < actor.length;i++){
        if (actor[i].moves.capa == "feu" | actor[i].moves.IA == "maintien")actor[i].moves.spam += 1;
        if (actor[i].moves.capa == "explode" && actor[i].moves.explode == 1){
            actor[i].moves.res = [1,2,3,4,5,6,7,8,9,10];
            actor[i].moves.vit = 0;
            actor[i].moves.explode = t;
        }
        else if (actor[i].moves.capa == "explode" && actor[i].moves.explode > 0 && t - actor[i].moves.explode > 2000){
            actor[i].moves.capa = "explosive";
            actor[i].moves.att.push(10);
            actor[i].img.src = actor[i].img.src = "images/" + actor[i].moves.img + "3.png";
            actor.forEach(
                function(c,index) {
                    contact(i,index);
                }
            );
        }
        else if (actor[i].moves.capa == "explosive" && t - actor[i].moves.explode > 2200)coup(i,15);
        else{
            if (actor[i].moves.capa == "instable"){
                actor[i].vx += actor[i].moves.inertie;
                if (actor[i].moves.inertie > 0)actor[i].moves.inertie -= 0.001;
                if (actor[i].moves.inertie < 0)actor[i].moves.inertie += 0.001;
                actor[i].moves.r += actor[i].moves.inertie / actor[i].moves.sy;
                if (actor[i].moves.r > Math.PI)actor[i].moves.r -= 2*Math.PI;
                if (actor[i].moves.r < Math.PI)actor[i].moves.r += 2*Math.PI;
            }
            if (tombe(i)) {
                actor[i].g += 1;
                if (actor[i].moves.capa == "immateriel") actor[i].g -= 0.9;
            }
            if (i != j){
                IA(i);
                contact(j,i);
            }
            if (actor[i].vx > 0) droite(i);
            else if (actor[i].vx < 0) gauche(i);
            actor[i].vy += actor[i].g;
            actor[i].x += actor[i].vx;
            actor[i].y += actor[i].vy;
            if (actor[i].moves.capa != "courseLongue"){
                actor[i].vx = 0;
            }
            actor[i].vy = 0;
        }
    }
    X = actor[j].x - 300;
    Y = actor[j].y - H / 2;
    draw();
    cible = proche();
    ctx.lineWidth = filin;
    ctx.beginPath();
    ctx.moveTo(actor[j].x - X,actor[j].y  - actor[j].moves.sy / 2 - Y);
    ctx.lineTo(actor[cible].x - X,actor[cible].y - actor[cible].moves.sy / 2 - Y);
    ctx.closePath();
    ctx.stroke();
}

function draw() {
    ctx.drawImage(imgFond,0,0);
    ctx.fillStyle = "rgb(45,150,0)";
    element.panneau.forEach(
        function(e) {
            ctx.drawImage(imgPancarte,e[0] - X - 30,e[1] - 70 - Y);
        }
    );
    element.balle.forEach(
        function(e) {
            ctx.drawImage(imgBalle,e[0] - X,e[1] - Y);
            if (actor[j].y - actor[j].moves.sy < e[1] + imgBalleY / 2 && actor[j].y > e[1] + imgBalleY / 2 && actor[j].x - actor[j].moves.sx / 2 < e[0] + imgBalleX / 2 && actor[j].x + actor[j].moves.sx / 2 > e[0] + imgFeuX / 2){
                element.balle.splice(element.balle.indexOf(e),1);
                balles += 1;
            }
        }
    );
    actor.forEach(
        function(c) {
            c.img.onload = function (){
                console.log("coucou");
            };
            if (c.moves.capa == "instable"){
                ctx.save();
                ctx.translate(c.x - X,c.y - c.moves.sy / 2 - Y);
                ctx.rotate(c.moves.r);
                ctx.drawImage(c.img,- c.moves.sx / 2,- c.moves.sy / 2);
                ctx.restore();
            }
            else ctx.drawImage(c.img,c.x - c.moves.sx / 2 - X,c.y - c.moves.sy - Y);
        }
    );
    niveau.forEach(
        function(c) {
            ctx.fillRect(c[0] - X,c[1] - Y,c[2],c[3]);
        }
    );
    element.feu.forEach(
        function(e) {
            ctx.drawImage(imgFeu,e[0] - X,e[1] - Y);
            e[0] += e[2];
            niveau.forEach(
                function(c) {
                    if (c[1] - 1 < e[1] + imgFeuY / 2 && c[1] + c[3] > e[1] + imgFeuY / 2 && c[0] < e[0] + imgFeuX / 2 && c[0] + c[2] > e[0] + imgFeuX / 2){
                        element.feu.splice(element.feu.indexOf(e),1);
                    }
                }
            );
            actor.forEach(
                function(c,index) {
                    if (c.y - c.moves.sy < e[1] + imgFeuY / 2 && c.y > e[1] + imgFeuY / 2 && c.x - c.moves.sx / 2 < e[0] + imgFeuX / 2 && c.x + c.moves.sx / 2 > e[0] + imgFeuX / 2){
                        element.feu.splice(element.feu.indexOf(e),1);
                        coup(index,3);
                    }
                }
            );
        }
    );
    if (laserPsy > -1 && Math.abs(actor[j].y - actor[laserPsy].y) < 500) {
        ctx.strokeStyle = "rgb(100,0,100)";
        ctx.lineWidth = 5;
        ctx.beginPath();
        ctx.moveTo(actor[j].x - X,actor[j].y  - actor[j].moves.sy / 2 - Y);
        ctx.lineTo(actor[laserPsy].x - X,actor[laserPsy].y - actor[laserPsy].moves.sy / 2 - Y);
        ctx.closePath();
        ctx.stroke();
    }
}

function drawDeath() {
    actor[j].g -= 1;
    actor[j].y -= actor[j].g;
    draw();
    window.requestAnimationFrame(drawDeath);
}

function drawTransfert() {
    if (((actor[vj].x - actor[j].x > 0 && X >= actor[vj].x - 300) | (actor[vj].x - actor[j].x < 0 && X <= actor[vj].x - 300)) | ((actor[vj].y - actor[j].y > 0 && Y >= actor[vj].y - H / 2) | (actor[vj].y - actor[j].y < 0 && Y <= actor[vj].y - H / 2))){
        j = vj;
        trans = 0;
        animation();
    }
    else{
        X += Math.round((actor[vj].x - actor[j].x) / 30) ;
        Y += Math.round((actor[vj].y - actor[j].y) / 30) ;
        draw();
        window.requestAnimationFrame(drawTransfert);
    }
}
